src/types.ts

```diff
@@ -195,3 +196,22 @@ export type GetDashboardCogMenuItemsRequest = RequireKeys<
 
 export interface GetDashboardCogMenuItemsResponse
     extends MenuItemRequestResponse {}
+
+export interface CustomAuthState {
+    args?: any;
+    isAuthenticated: boolean;
+}
+export interface AuthState extends CustomAuthState {
+    status: AuthStateStatus;
+}
+
+export interface AuthStateOptions {
+    url: string;
+    closePopupAfterAuth?: boolean;
+    authStateCallback: () =>
+        | Promise<CustomAuthState | boolean>
+        | CustomAuthState
+        | boolean;
+    retryInterval?: number;
+    totalTimeout?: number;
+}
```

Comment:
I think we should obscure some of these options from the user unless users explicitly ask for control over these things. Doing so will give us more flexibility over the implementation details while we figure things out. I'd say we can remove `retryInterval`, `totalTimeout`, `closePopupAfterAuth` and keep the other two

Reply:
They are optional with defaults so they are there if needed. I think some power users would want to fine tune the experience depending on the how expensive/performance the underlying logic in their `authStateCallback`. At least one customer already requested to close the popup themselves so I think providing the flexibly without forcing it is good.

Reply:
I'm explicitly thinking about what will happen if we do implement a way for the developer to tell us that the auth state is resolved without polling, like my redirect idea or a different method. In that case `retryInterval` and `totalTimeout` might not apply anymore. `closePopupAfterAuth` won't apply if we choose a different auth method besides a popup. Keep in mind we can always give them control over those settings if they ask for it ... going the opposite way is harder